internal/util.go

```diff
@@ -12,13 +12,17 @@ import (
 // for resources that can contain an arbitrary or dynamic number of fields.
 func RemainingKeys(s interface{}, m map[string]interface{}) (extras map[string]interface{}) {
 	extras = make(map[string]interface{})
+	for k, v := range m {
+		extras[k] = v
+	}
+
 	valueOf := reflect.ValueOf(s)
 	typeOf := reflect.TypeOf(s)
 	for i := 0; i < valueOf.NumField(); i++ {
 		field := typeOf.Field(i)
 		tagValue := field.Tag.Get("json")
-		if _, ok := m[tagValue]; !ok {
-			extras[tagValue] = m[tagValue]
+		if _, ok := extras[tagValue]; ok {
+			delete(extras, tagValue)
```

Comment:
I'm sorry, but I'm not following. :(

To start from scratch: From my point of view, the objective of this function is to find extra fields that were sent in a JSON response. For example:

```json
{
  "user_id": "abcd1234",
  "username": "John",
  "custom_field": "foo"
}
```

This response needs to be unmarshaled into into a struct, but it's impossible to have a definitive struct schema for an arbitrary amount of unknown fields. So we have a struct like this:

```go
type User struct {
  UserID string
  Username string
  Extra map[string]interface{}
}
```

Because Go's style of naming fields can differ from the JSON response, and to ensure the correct JSON field maps to the struct field, JSON tags are added:

```go
type User struct {
  UserID string `json:"user_id"`
  Username string `json:"username"`
  Extra map[string]interface{}
}
```

All of Gophercloud's results contain the correct JSON tags, effectively creating a confident bridge between the struct field and the JSON response.

So that's why I chose to use a method of reduction. My approach was to make a `map[string]interface{}` out of the JSON response, which effectively creates a simple map where the keys are the field in the JSON response. The known keys can be confidently referenced to a result struct, which includes mappings like `UserID` to `user_id`.

Anything that can't be mapped can be considered unknown to Gophercloud and thus be considered an "extra" field.

I'm having trouble trying to apply your suggested way to this method. If I were to just use the struct fields, that would cause problems with names such as `UserID`.